parsec/core/fs/workspacefs/versioning_helpers.py

```diff
@@ -50,109 +51,187 @@ class ManifestDataAndPaths(NamedTuple):
                 )
         return manifest
 
-    async def _list_versions(entry_id: EntryID):
-        if entry_id in versions_list_cache:
-            return versions_list_cache[entry_id]
-        versions_list_cache[entry_id] = await workspacefs.remote_loader.list_versions(entry_id)
-        return versions_list_cache[entry_id]
 
-    async def _get_past_path(entry_id: EntryID, version=None, timestamp=None) -> FsPath:
+class VersionsListCache:
+    """
+    Caches results of the remote_loader.list_versions calls (working on EntryIDs).
+    """
+
+    def __init__(self, remote_loader):
+        self._versions_list_cache = {}
+        self._remote_loader = remote_loader
+
+    async def load(self, entry_id: EntryID):
+        """
+        Raises:
+            FSError
+            FSBackendOfflineError
+            FSWorkspaceInMaintenance
+            FSRemoteManifestNotFound
+        """
+        if entry_id in self._versions_list_cache:
+            return self._versions_list_cache[entry_id]
+        self._versions_list_cache[entry_id] = await self._remote_loader.list_versions(entry_id)
+        return self._versions_list_cache[entry_id]
+
 
-        # Get first manifest
+async def _get_path_at_timestamp(
+    manifest_cache: ManifestCache, entry_id: EntryID, timestamp: Pendulum
+) -> FsPath:
+    """
+    Find a path for an entry_id at a specific timestamp.
+
+    If the path is broken, will raise an EntryNotFound exception. All the other exceptions are
+    thrown by the ManifestCache.
+
+    Raises:
+        FSError
+        FSBackendOfflineError
+        FSWorkspaceInMaintenance
+        FSBadEncryptionRevision
+        FSWorkspaceNoAccess
+        EntryNotFound
+    """
+    # Get first manifest
+    try:
+        current_id = entry_id
+        current_manifest = await manifest_cache.load(current_id, timestamp=timestamp)
+    except FSRemoteManifestNotFound:
+        raise EntryNotFound(entry_id)
+
+    # Loop over parts
+    parts = []
+    while not is_workspace_manifest(current_manifest):
+
+        # Get the manifest
         try:
-            current_id = entry_id
-            current_manifest = await _load_manifest_or_cached(
-                current_id, version=version, timestamp=timestamp
+            parent_manifest = await manifest_cache.load(
+                current_manifest.parent, timestamp=timestamp
             )
-        except FSLocalMissError:
+        except FSRemoteManifestNotFound:
             raise EntryNotFound(entry_id)
 
-        # Loop over parts
-        parts = []
-        while not is_workspace_manifest(current_manifest):
+        # Find the child name
+        try:
+            name = next(
+                name
+                for name, child_id in parent_manifest.children.items()
+                if child_id == current_id
+            )
+        except StopIteration:
+            raise EntryNotFound(entry_id)
+        else:
+            parts.append(name)
```

Comment:
This part could be written with a `for-else`, although it's a matter of taste:
```python
# Find the child name
for name, child_id in parent_manifest.children.items():
    if child_id == current_id:
        parts.append(name)
        break
else:
    raise EntryNotFound(entry_id)
```

Reply:
Changed it to yours as I find it cleaner too